src/common/faucet.ts

```diff
@@ -0,0 +1,234 @@
+/**
+ * Copyright (C) 2022-2024 Permanent Data Solutions, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+import { ARIOWithFaucet, TokenFaucet } from '../types/faucet.js';
+import { ARIOReadable, ARIOWriteable } from './io.js';
+
+const DEFAULT_FAUCET_API_URL = 'https://faucet.ario.permaweb.services';
+
+/**
+ * Creates a proxy object that implements the TokenFaucet interface. It wraps the ARIOReadable instance and adds methods for claiming tokens from the faucet API.
+ * @param arioInstance - The ARIOReadable instance
+ * @param faucetApiUrl - The URL of the faucet API
+ * @returns A proxy object that implements the TokenFaucet interface
+ */
+export function createFaucet({
+  arioInstance,
+  faucetApiUrl = DEFAULT_FAUCET_API_URL,
+}: {
+  arioInstance: ARIOReadable | ARIOWriteable;
+  faucetApiUrl?: string;
+}): ARIOWithFaucet<ARIOReadable | ARIOWriteable> {
+  const faucet = new ARIOTokenFaucet({
+    faucetUrl: faucetApiUrl,
+    processId: arioInstance.process.processId,
+  });
+
+  const proxy = new Proxy(arioInstance, {
+    get(target, prop) {
+      if (prop === 'faucet') {
+        return faucet;
+      }
+      if (prop in target) {
+        const result = target[prop as keyof typeof target];
+        if (typeof result === 'function') {
+          return result.bind(target);
+        }
+        return result;
+      }
+      return undefined;
+    },
+  });
+  return proxy as ARIOWithFaucet<ARIOReadable | ARIOWriteable>;
+}
+
+export class ARIOTokenFaucet implements TokenFaucet {
+  private faucetUrl: string;
+  private processId: string;
+
+  constructor({
+    faucetUrl,
+    processId,
+  }: {
+    faucetUrl: string;
+    processId: string;
+  }) {
+    this.faucetUrl = faucetUrl;
+    this.processId = processId;
+  }
+
+  /**
+   * Returns the captcha URL for a process. The captcha is used to verify a human is solving the captcha. Once you have a captcha response, you can use it to request an authorization token via the requestAuthToken method.
+   * @returns The captcha URL for a process
+   */
+  async captchaUrl(): Promise<{
+    processId: string;
+    captchaUrl: string;
+  }> {
+    const res = await fetch(
+      `${this.faucetUrl}/api/captcha/url?process-id=${this.processId}`,
+      {
+        method: 'GET',
+      },
+    );
+
+    if (!res.ok) {
+      const body = await res.json().catch(() => ({ error: res.statusText }));
+      throw new Error((body as { error: string }).error);
+    }
```

Comment:
nit: inconsistent `async` + `await` vs `then` + `catch`. staying consistent improves readability and stack traces. not sure what the error outputs here when `await fetch(..` and `await.json()...` but a try/catch around the whole function may improve DX in unhappy paths